python/dgl/batched_heterograph.py

```diff
@@ -0,0 +1,410 @@
+from .frame import FrameRef, Frame
+from .heterograph import DGLHeteroGraph
+
+__all__ = ['BatchedDGLHeteroGraph', 'unbatch_hetero', 'batch_hetero']
+
+class BatchedDGLHeteroGraph(DGLHeteroGraph):
+    """Class for batched DGLHeteroGraphs.
+
+    A :class:`BatchedDGLHeteroGraph` basically merges a list of small graphs into a giant
+    graph so that one can perform message passing and readout over a batch of graphs
+    simultaneously.
+
+    For a given node/edge type, the nodes/edges are re-indexed with a new id in the
+    batched graph with the rule below:
+
+    ======  ==========  ========================  ===  ==========================
+    item    Graph 1     Graph 2                   ...  Graph k
+    ======  ==========  ========================  ===  ==========================
+    raw id  0, ..., N1       0, ..., N2           ...  ..., Nk
+    new id  0, ..., N1  N1 + 1, ..., N1 + N2 + 1  ...  ..., N1 + ... + Nk + k - 1
+    ======  ==========  ========================  ===  ==========================
+
+    To modify the features in :class:`BatchedDGLHeteroGraph` has no effect on the original
+    graphs. See the examples below about how to work around.
+
+    Parameters
+    ----------
+    graph_list : iterable
+        A collection of :class:`~dgl.DGLHeteroGraph` to be batched.
+    node_attrs : None or dict
+        The node attributes to be batched. If ``None``, the resulted graph will not have
+        features. If ``dict``, it maps str to str or iterable. The keys represent names of
+        node types and the values represent the node features to be batched for the
+        corresponding type. By default, we use all features for all types of nodes.
+    edge_attrs : None or dict
+        Same as for the case of :attr:`node_attrs`.
+
+    Examples
+    --------
+    Create two :class:`~dgl.DGLHeteroGraph` objects.
+
+    **Instantiation:**
+
+    >>> import dgl
+    >>> import torch as th
+    >>> g1 = dgl.heterograph({
+    ...    ('user', 'follows', 'user'): [(0, 1), (1, 2)],
+    ...    ('user', 'plays', 'game'): [(0, 0), (1, 0)]
+    ... })
+    >>> g1.nodes['user'].data['h1'] = th.tensor([[0.], [1.], [2.]])
+    >>> g1.nodes['user'].data['h2'] = th.tensor([[3.], [4.], [5.]])
+    >>> g1.nodes['game'].data['h1'] = th.tensor([[0.]])
+    >>> g1.edges['plays'].data['h1'] = th.tensor([[0.], [1.]])
+
+    >>> g2 = dgl.heterograph({
+    ...    ('user', 'follows', 'user'): [(0, 1), (1, 2)],
+    ...    ('user', 'plays', 'game'): [(0, 0), (1, 0)]
+    ... })
+    >>> g2.nodes['user'].data['h1'] = th.tensor([[0.], [1.], [2.]])
+    >>> g2.nodes['user'].data['h2'] = th.tensor([[3.], [4.], [5.]])
+    >>> g2.nodes['game'].data['h1'] = th.tensor([[0.]])
+    >>> g2.edges['plays'].data['h1'] = th.tensor([[0.], [1.]])
+
+    Merge two :class:`~dgl.DGLHeteroGraph` objects into one :class:`BatchedDGLHeteroGraph` object.
+    When merging a list of graphs, we can choose to include only a subset of the attributes.
```

Comment:
It's better to give the simplest example first such as just `bg = dgl.batch_hetero([g1, g2])`.

Reply:
Better?